src/rgw/rgw_sal_d3n.cc

```diff
@@ -0,0 +1,303 @@
+    r_obj.oid = oid_in_cache;
+    ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): READ FROM CACHE: oid=" << oid_in_cache << " length to read is: " << len_to_read << " part num: " << start_part_num << " read_ofs: " << read_ofs << " part len: " << part_len << dendl;
+    if (filter->get_d3n_cache()->get(oid_in_cache, part_len)) {
+      // Read From Cache
+      auto completed = aio->get(r_obj, rgw::Aio::d3n_cache_op(dpp, y, read_ofs, len_to_read, filter->get_d3n_cache()->cache_location), cost, id);
+      ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Info: flushing data for oid: " << oid_in_cache << dendl;
+      auto r = flush(dpp, std::move(completed));
+      if (r < 0) {
+        ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Error: failed to flush, r= " << r << dendl;
+        return r;
+      }
+    } else {
+      //for ranged requests, for last part, the whole part might exist in the cache
+      oid_in_cache = source->get_bucket()->get_marker() + "_" + oid + "_" + std::to_string(adjusted_start_ofs) + "_" + std::to_string(obj_max_req_size);
+      r_obj.oid = oid_in_cache;
+      ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): READ FROM CACHE: oid=" << oid_in_cache << " length to read is: " << len_to_read << " part num: " << start_part_num << " read_ofs: " << read_ofs << " part len: " << part_len << dendl;
+      if (filter->get_d3n_cache()->get(oid_in_cache, obj_max_req_size)) {
+        // Read From Cache
+        auto completed = aio->get(r_obj, rgw::Aio::d3n_cache_op(dpp, y, read_ofs, len_to_read, filter->get_d3n_cache()->cache_location), cost, id);
+        ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Info: flushing data for oid: " << oid_in_cache << dendl;
+        auto r = flush(dpp, std::move(completed));
+        if (r < 0) {
+          ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Error: failed to flush, r= " << r << dendl;
+          return r;
+        }
+      } else {
+        ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Info: draining data for oid: " << oid_in_cache << dendl;
+        auto r = drain(dpp);
+        if (r < 0) {
+          ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Error: failed to drain, r= " << r << dendl;
+          return r;
+        }
+        break;
+      }
+    }
+    if (start_part_num == (num_parts - 1)) {
+      ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Info: draining datafor oid: " << oid_in_cache << dendl;
+      return drain(dpp);
+    } else {
+      adjusted_start_ofs += obj_max_req_size;
+    }
+    start_part_num += 1;
+    len -= obj_max_req_size;
+  } while(start_part_num < num_parts);
+
+  ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Fetching object from backend store" << dendl;
+  if (source->is_compressed()) {
```

Comment:
@cbodley: how can i find if the object is encrypted? using xattrs?

Reply:
yeah, RGW_ATTR_CRYPT_MODE